fog/ledger/server/src/db_fetcher.rs

```diff
@@ -0,0 +1,299 @@
+        if block_index == 0 {
+            return u64::MAX;
+        }
+
+        // Timer that tracks how long we have had WatcherBehind error for,
+        // if this exceeds watcher_timeout, we log a warning.
+        let mut watcher_behind_timer = Instant::now();
+        loop {
+            match watcher.get_block_timestamp(block_index) {
+                Ok((ts, res)) => match res {
+                    TimestampResultCode::WatcherBehind => {
+                        if watcher_behind_timer.elapsed() > watcher_timeout {
+                            log::warn!(logger, "watcher is still behind on block index = {} after waiting {} seconds, ledger service will be blocked", block_index, watcher_timeout.as_secs());
+                            watcher_behind_timer = Instant::now();
+                        }
+                        std::thread::sleep(Self::POLLING_FREQUENCY);
+                    }
+                    TimestampResultCode::BlockIndexOutOfBounds => {
+                        log::warn!(logger, "block index {} was out of bounds, we should not be scanning it, we will have junk timestamps for it", block_index);
+                        return u64::MAX;
+                    }
+                    TimestampResultCode::Unavailable => {
+                        log::crit!(logger, "watcher configuration is wrong and timestamps will not be available with this configuration. Ledger service is blocked at block index {}", block_index);
+                        std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                    }
+                    TimestampResultCode::WatcherDatabaseError => {
+                        log::crit!(logger, "The watcher database has an error which prevents us from getting timestamps. Ledger service is blocked at block index {}", block_index);
+                        std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                    }
+                    TimestampResultCode::TimestampFound => {
+                        return ts;
+                    }
+                },
+                Err(err) => {
+                    log::error!(
+                        logger,
+                        "Could not obtain timestamp for block {} due to error {}, this may mean the watcher is not correctly configured. will retry",
+                        block_index,
+                        err
+                    );
+                    std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                }
+            };
+        }
+    }
+
+    fn add_records_to_enclave(&mut self, block_index: u64, records: Vec<KeyImageData>) {
+        let num_records = records.len();
+
+        let _info = retry(delay::Fixed::from_millis(5000), || {
+            trace_time!(
+                self.logger,
+                "Added {} records into the enclave",
+                num_records
+            );
+            let _metrics_timer = counters::ENCLAVE_ADD_KEY_IMAGE_DATA_TIME.start_timer();
```

Comment:
Not a huge problem, but this has the undesirable effect of updating the metrics even if the operation has failed. There's an easy solution for that: If you call `.stop_and_discard()` on the returned value, it would skip the collection. See https://github.com/mobilecoinfoundation/fog/blob/78386dfcd41b9b77f68350eee915a989a3d78533/fog/ingest/server/src/controller.rs#L533 for example

Reply:
done